### PR TITLE
Run unit tests in parallel by default. 

### PR DESCRIPTION
**Pull request by jonathanknowles**
This PR adjusts the following unit test suites to run tests in parallel by default:
- `cardano-coin-selection:test:unit`
- `cardano-wallet:test:unit`
- `cardano-wallet-primitive:test:unit`
- `strict-non-empty-containers:test:unit`

It uses the `SpecHook` feature of `hspec`, as described here: https://hspec.github.io/parallel-spec-execution.html 

This makes it unnecessary to manually use the  `parallel` combinator for every individual `Spec`.

For the `cardano-wallet:test:unit` test suite in particular, this leads to a large speedup when run on a machine with at least 4 cores:

### Original timings
```
Finished in 705.3929 seconds, used 1913.9534 seconds of CPU time
real    11m49.908s
user    25m45.315s
sys     6m17.046s
```

### After using `SpecHook` to run tests in parallel by default
```
Finished in 500.0908 seconds, used 1755.7065 seconds of CPU time
real    8m25.017s
user    25m7.003s
sys     4m17.505s
```

### After removing all manual calls to `parallel` (no significant slowdown or speedup)
```
Finished in 506.2987 seconds, used 1779.8373 seconds of CPU time
real    8m31.305s
user    25m23.259s
sys     4m25.264s
```

### CODE REVIEW

**Review by paolino — APPROVED** [review]
Thanks, let's go (in parallel)

**Comment by jonathanknowles** [comment]
I noticed the following failure on Buildkite https://buildkite.com/input-output-hk/cardano-wallet/builds/24598#01877ef6-c97a-4499-a502-21729ce123b6/6-2783:

```
cardano-wallet-test-unit> Failures:
cardano-wallet-test-unit>   test/unit/Cardano/Wallet/Shelley/NetworkSpec.hs:196:9:
cardano-wallet-test-unit>   1) Cardano.Wallet.Shelley.Network.Observer, typical use, when refresh fails, only MsgWillFetch is traced
cardano-wallet-test-unit>        expected: [MsgWillFetch (fromList ["k"])]
cardano-wallet-test-unit>         but got: [MsgWillFetch (fromList ["k"]),MsgRemovedObserver "k",MsgWillFetch (fromList []),MsgDidFetch (fromList [])]
cardano-wallet-test-unit>   To rerun use: --match "/Cardano.Wallet.Shelley.Network/Observer/typical use/when refresh fails/only MsgWillFetch is traced/"
```

Though it succeeded the second time round: https://buildkite.com/input-output-hk/cardano-wallet/builds/24598#01877f13-bb8b-41e3-9856-d954602286c8

I'll see if I can reproduce locally.

UPDATE: It turns out that `observerSpec` needs to be run sequentially, as consecutive tests rely on shared state. I've fixed this in ef798a87e20edfe0a8a26689b51d4c2b54040803.

File: lib/primitive/test/spec/Cardano/Wallet/Primitive/CollateralSpec.hs [hunk 1]
```diff
@@ -669,7 +669,7 @@ instance Arbitrary TxOut where
 
 spec :: Spec
 spec = do
-    parallel $ do
+    do
```
Comment (Unisay) [inline lib/primitive/test/spec/Cardano/Wallet/Primitive/CollateralSpec.hs]: Mb one `do` suffice? :)

File: lib/coin-selection/test/spec/SpecHook.hs [hunk 2]
```diff
@@ -0,0 +1,6 @@
+module SpecHook where
+
+import Test.Hspec
+
+hook :: Spec -> Spec
+hook = parallel
```
Comment (Unisay) [inline lib/coin-selection/test/spec/SpecHook.hs:6]: 👍🏼

File: lib/primitive/test/spec/Cardano/Wallet/Primitive/CollateralSpec.hs (hunk 1 above)
Comment (paolino) [reply lib/primitive/test/spec/Cardano/Wallet/Primitive/CollateralSpec.hs]: parallel do

Comment (jonathanknowles) [reply lib/primitive/test/spec/Cardano/Wallet/Primitive/CollateralSpec.hs]: > Mb one `do` suffice? :)

Nice find! 😆

File: lib/wallet/test/unit/SpecHook.hs [hunk 3]
```diff
@@ -0,0 +1,6 @@
+module SpecHook where
```
Comment (Unisay) [inline lib/wallet/test/unit/SpecHook.hs:1]: I've noticed there is already `lib/test-utils/test/Test/Hspec/ExtraSpec.hs` for a similar purposes, do you think it makes sense to move this utility there?

File: lib/primitive/test/spec/Cardano/Wallet/Primitive/CollateralSpec.hs (hunk 1 above)
Comment (jonathanknowles) [reply lib/primitive/test/spec/Cardano/Wallet/Primitive/CollateralSpec.hs]: Fixed in 3287b9926c9e4ad517a8e47f09d20e60625455e5.

File: lib/wallet/test/unit/SpecHook.hs (hunk 3 above)
Comment (jonathanknowles) [reply lib/wallet/test/unit/SpecHook.hs:1]: I suspect that module is to help with integration tests. So perhaps we could leave it untouched for the moment.

**Pull request merged**